src/coreclr/vm/jitinterface.cpp

```diff
@@ -6352,6 +6362,49 @@ CORINFO_CLASS_HANDLE  CEEInfo::getTypeForBox(CORINFO_CLASS_HANDLE  cls)
     return static_cast<CORINFO_CLASS_HANDLE>(VMClsHnd.AsPtr());
 }
 
+/***********************************************************************/
+// Get a representation for a stack-allocated boxed value type
+//
+CORINFO_CLASS_HANDLE  CEEInfo::getTypeForBoxOnStack(CORINFO_CLASS_HANDLE cls)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    CORINFO_CLASS_HANDLE result = NULL;
+
+    JIT_TO_EE_TRANSITION();
+
+    TypeHandle VMClsHnd(cls);
+    if (Nullable::IsNullableType(VMClsHnd)) 
+    {
+        VMClsHnd = VMClsHnd.AsMethodTable()->GetInstantiation()[0];
+    }
+
+#ifdef FEATURE_64BIT_ALIGNMENT
+    if (VMClsHnd.RequiresAlign8())
+    {
+        // TODO: Maybe handle 32 bit platforms with 8 byte alignments
+        result = NULL;
+    }
+    else
+#endif
+    {
+        TypeHandle mt(CoreLibBinder::GetElementType(ELEMENT_TYPE_I));
+        TypeHandle boxedFields[2] = {mt, VMClsHnd};
+        Instantiation boxedFieldsInst((TypeHandle*)&boxedFields, 2);
+        TypeHandle genericKvp = CoreLibBinder::GetClass(CLASS__KEYVALUEPAIRGENERIC);
```

Comment:
This won't work well. KeyValuePair is auto-layout when the type contains references. auto-layout moves references to the front.

For example, `KeyValuePair<IntPtr, string>` has string at offset 0 and IntPtr at offset 8.

Reply:
I guess it probably works ok for valuetypes currently, but it is fragile.

Reply:
Thanks... we were trying to find a suitable type without needing too much from the runtime.

Reply:
I think we should introduce dedicated internal helper type for this.

Reply:
We can also try and do this entirely within the jit (that's where I started), but it would be nice to have class and field handles and the like.

I also am still not sure how effective any of this will be, so I want to do some more analysis. Earlier studies suggest we can prove maybe 1% of boxes do not escape (static, not dynamic) but were based on SPMI which can't replay a bunch of possibly interesting cases. Given the cost of the analysis I would like to see if we can get that number a bit higher.

Reply:
That 1% is a static measure, that is, 1% of the box call sites, not necessarily 1% of boxes allocated.